pkg/authentication/middlewares.go

```diff
@@ -38,29 +48,82 @@ func (m *Middleware) isAllowListed(r *http.Request) bool {
 	return false
 }
 
-func (m *Middleware) OAuth2Authentication(next http.Handler) http.Handler {
+func (m *Middleware) OAuth2AuthenticationChain() []func(http.Handler) http.Handler {
+	return []func(http.Handler) http.Handler{
+		m.OAuth2BearerAuthentication,
+		m.OAuth2CookieAuthentication,
+	}
+}
+
+func (m *Middleware) OAuth2BearerAuthentication(next http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		ctx, span := m.tracer.Start(r.Context(), "authentication.Middleware.OAuth2Authentication")
+		ctx, span := m.tracer.Start(r.Context(), "authentication.Middleware.OAuth2BearerAuthentication")
 		defer span.End()
-		r = r.WithContext(ctx)
 
 		if m.isAllowListed(r) {
-			next.ServeHTTP(w, r.WithContext(ctx))
+			next.ServeHTTP(w, r)
 			return
 		}
 
-		rawAccessToken, err := getBearerToken(r.Header)
-		if err != nil {
-			m.unauthorizedResponse(w, err)
-			return
-		}
+		var (
+			principal *Principal
+			err       error
+		)
 
 		// add the Otel HTTP Client
 		r.WithContext(OtelHTTPClientContext(r.Context()))
 
-		principal, err := m.oauth2.Verifier().VerifyAccessToken(r.Context(), rawAccessToken)
+		if rawAccessToken, found := m.getBearerToken(r.Header); found {
+			principal, err = m.oauth2.Verifier().VerifyAccessToken(r.Context(), rawAccessToken)
+			if err != nil {
+				m.unauthorizedResponse(w, err)
+				return
+			}
+		}
+
+		ctx = PrincipalContext(ctx, principal)
+		next.ServeHTTP(w, r.WithContext(ctx))
+	})
+}
+
+func (m *Middleware) OAuth2CookieAuthentication(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx, span := m.tracer.Start(r.Context(), "authentication.Middleware.OAuth2CookieAuthentication")
+		defer span.End()
+
+		if m.isAllowListed(r) {
+			next.ServeHTTP(w, r)
+			return
+		}
+
+		err := fmt.Errorf("no authentication token found")
+
+		principal := PrincipalFromContext(r.Context())
+		if principal != nil {
+			// principal != nil means bearer authentication set the principal and we can move on
+			next.ServeHTTP(w, r)
+			return
+		}
```

Comment:
Middlewares are ordered in nature, they are intended to be registered in the desired order based on their processing function. If you take a look at other frameworks, they all state that _middlewares order matters_ ([Express.js](https://expressjs.com/en/guide/writing-middleware.html#:~:text=The%20order%20of%20middleware%20loading,first%20are%20also%20executed%20first.), [Django](https://docs.djangoproject.com/en/5.0/topics/http/middleware/#:~:text=The%20order%20in%20MIDDLEWARE%20matters%20because%20a%20middleware%20can%20depend%20on%20other%20middleware), [Java](https://www.oracle.com/java/technologies/filters.html#:~:text=The%20order%20of%20the%20filters%20in%20the%20chain%20is%20the%20same%20as%20the%20order%20that%20filter%20mappings%20appear%20in%20the%20web%20application%20deployment%20descriptor.)), so it's natural to decompose processing in multiple steps.
The alternative would be to have all in one middleware like it was before my previous refactoring, but that has been excluded.

Also, note that the function used to register them provides the two middlewares in order, so there is no way to accidentally register them in the wrong order (unless one explicitly wants to do so of course, but that's criminal behavior 😃).
What I can do is not export them, since right now they are (but it's not needed), so I'll change their name to a lowercase so they can't be used withtout the providing function `OAuth2AuthenticationChain`.